File: day 12/hillClimbing.py


class HillClimb:

    def __init__(self, filename: str) -> None:
        self.grid = []
        self.grid2 = []
        with open(filename) as f:
            for i, line in enumerate(f.read().splitlines()):
                if line.count('S'):
                    self.start = (i, line.find('S'))
                    line = line.replace('S', 'a')
                if line.count('E'):
                    self.end = (i, line.find('E'))
                    line = line.replace('E', 'z')
                self.grid.append([ord(letter) for letter in list(line)])
                self.grid2.append(list(line))
        self.nrow = len(self.grid)
        self.ncol = len(self.grid[0])

    def BFS(self):
        self.visited = {self.start: self.start}
        queue = [self.start]
        while len(queue) and queue[0] != self.end:
            ic, jc = queue.pop(0)
            for i, j in ([(ic, jc-1), (ic, jc+1), (ic-1, jc), (ic+1, jc)]):
                if (i, j) in self.visited:
                    continue
                if not(0 <= i < self.nrow and 0 <= j < self.ncol):
                    continue
                if abs(self.grid[i][j] - self.grid[ic][jc]) > 1:
                    continue
                    # visited.add((i, j))
                    # print(abs(self.grid[i][j] - self.grid[ic][jc]))
                queue.append((i, j))
                self.visited[(i, j)] = (ic, jc)
                print(self.grid2[i][j])
                    # print(self.grid2[i][j])
                    # print((i, j) == self.end)
                if (i, j) == self.end:
                    return (i, j)
        # return self.start
    
    def shortest_path(self) -> int:
        i, j = self.BFS()
        # print(len(self.visited))
        path = [(i, j)]
        while (i, j) != self.start:
            (i, j) = self.visited[(i, j)]
            path.append((i, j))
        return len(path)-1

if __name__ == "__main__":
    hc = HillClimb("day 12/simple.txt")
    print(hc.shortest_path())
    

File: day 12/hillClimbing.cpp
#include <string>
#include <iostream>
#include <fstream>
#include <vector>
#include <queue>
#include <unordered_map>
#include <unordered_set>

using IntPair = std::pair<int, int>;

namespace std {
	template <>
		struct hash<IntPair> {
			std::size_t operator() (const IntPair& p) const {
			std::string uniqueIntPairString = std::to_string(p.first) + "##" + std::to_string(p.second);
			std::hash<std::string> stringHasher;
			return stringHasher(uniqueIntPairString);
			}
	};
}

class HillClimb{
	std::vector<std::vector<int>> grid;
	std::unordered_map<IntPair, IntPair> visited;
	IntPair start;
	IntPair end;
	int nrow;
	int ncol;

public:
	HillClimb(const char* filename) 
	{
		std::ifstream file(filename);
		std::string line;
		int i = 0;
		while (file >> line) {
			auto j = line.find("S");
			if (j != std::string::npos)
			{
				end = std::make_pair(i, j);
				line.replace(j, 1, "a");
			}
			j = line.find("E");
			if (j != std::string::npos)
			{
				start = std::make_pair(i, j);
				line.replace(j, 1, "z");
			}
			++i;
			grid.emplace_back(line.begin(), line.end());
		}

		nrow = static_cast<int>(grid.size());
		ncol = static_cast<int>(grid.front().size());
	}

	IntPair BFS() {
		visited.emplace(start, start);
		std::queue<IntPair> q;
		q.emplace(start);
		while (!q.empty())
		{
			auto [ic, jc] = q.front();
			q.pop();
			for (auto [i, j]: neighbors(ic, jc))
			{
				if (visited.contains(std::make_pair(i, j)))
					continue;
				if ((grid[i][j] - grid[ic][jc]) < -1)
					continue;
				q.emplace(i, j);
				visited.emplace(std::make_pair(i, j), std::make_pair(ic, jc));
				if (grid[i][j] == static_cast<int>('a')) // Part B
					return std::make_pair(i, j);
				// if (std::make_pair(i, j) == end) //Part A
					// return std::make_pair(i, j);
			}
		}
		return start;
	}

	std::unordered_set<IntPair> neighbors(int ic, int jc)
	{
		std::unordered_set<IntPair> neighbors;
		neighbors.emplace(ic, std::max(jc-1, 0));
		neighbors.emplace(ic, std::min(jc+1, ncol-1));
		neighbors.emplace(std::max(ic-1, 0), jc);
		neighbors.emplace(std::min(ic+1, nrow-1), jc);
		return neighbors;
	}

	int shortest_path(){
		auto idx = BFS();
		std::vector<IntPair> path = {idx};
		while (visited.contains(idx) && idx != start){
			idx = visited.at(idx);
			path.emplace_back(idx);
		}
		return static_cast<int>(path.size())-1;
	}
};

int main(){

	HillClimb hc{"input.txt"};
	std::cout << hc.shortest_path() << std::endl;
	return 1;
}
	